@nrwl/angular:ngrx is not backward compatible or unique across store generations

SUBJECT: NgRX is not backward compatible or unique across store generations

This is now confirmed after double re-iterating thru a prior ticket was created but closed due to false false alarm. https://github.com/nrwl/nx/issues/10880

I don't want to strike any nerves with the nrwl committers... Im a java developer of 16yrs and typescript 7. Java has withstood the test of time for 30+yrs and the field continues to fallback on it. React has not. React is javascript centric. Javascript is extremely dangerous and unfit for commercial development. Its nice that @nrwl/nx has been kind enough to honor both react and angular and for this schematic we have classes and creators. And obviously Im a classes guy. Java folks make the best typescript developers hands down. I find the react javascript inclination is not readable (yes you can do alot of whizbang stuff like code your own html tag whereas in angular its more difficult react can do things angular cant). But I stand by nature and the natural world. Classes are better for the natural world. And I frown when I see react patterns in this framework start to take over its aggressive and domineering. Usability should be first and foremost for any framework thats why I committed to typescript its just like java. I highly encourage the development to stay the course. When Props were introduced Im like oh god here comes react... the quick fix the quick thing not even named right borderline reckless. And it's just not palatable. And again I defer back to Java and its 30yrs of tested time.

This schematic has taken on the form (And it's drifted) and inclination leaning more towards react and creators and I find passing around variables attached to functions and functional programming obscene fit for javascript functional programmers who know only one thing (not object oriented design which java is all about for 30yrs) and love the cryptic and arcane ... javascript incline is way too dangerous not even fun. There is a natural world in front of us. NATURE. The originators of NgRX nailed NgRX first time done right Just like RxJS (RxJS is 100% symmetrically fit for quantum physics RxJS will manage control and operate an angle (angel) of light... real atomic particle energy).  The functional programming tendencies are not fit to the natural world, if they were javascript would never have had to been hijacked, re-invented by typescript and re-transformed by Microsoft to save google.

There is a rivalry and collision with react devs and angular devs in the field a huge difference. BUT Angular leaves react completely in the dust for what it started out being and standardized with multiple levels angular2, ngrx and now nrwl/nx. I hope NgRX keeps to its original form classes being the dominant inclination and nature instead of the fast whz bang javascript functional stuff "creators".

Victor Savkin is the high priest of computer science on this planet. He has guaranteed that what will be produced will be capable of surviving change, scale to unlimited dimensions and fit to model the natural world. Few frameworks come close... I would say RxJS and NestJS.

Thanks for your promotion of an awesome library.


## Current Behavior
After defining a slew of stores using the @nrwl/angular:ngrx schematic (this will affect nxconsole) I have to go thru all the code and rename everything thats exported due to name collisions. This severely degrades development fluidity. This was never the case early on with earlier versions of NgRX maybe this schematic is fresh code.
Regardless, this area needs to be abstracted to survive real world development.


## Expected Behavior
If I generate stores using the @nrwl/angular:ngrx schematic it should guarantee the following:

1. that it pre-pends all particular exported members of the generated code with the name of the store to avoid subsequent naming collisions across other generated stores, their folders and exported members
2. This is suggested to violate backward compatibility but thats up for debate since there never was a schematic prior but I never had naming collisions issues with older versioned ngrx with angular ensuring hand coding uniqueness maybe it was not abstracted well enough.

Thanks for a terrific job this schematic is a dream

## Steps to Reproduce
see https://github.com/nrwl/nx/issues/10880

### Failure Logs
```
Building entry point '@meanstacknh/shared-root'
------------------------------------------------------------------------------
✖ Compiling with Angular sources in Ivy partial compilation mode.
libs/shared/root/data-access/src/index.ts:36:1 - error TS2308: Module './lib/+state/auth/auth.actions' has already exported a member named 'Update'. Consider explicitly re-exporting to resolve the ambiguity.

36 export * from './lib/+state/user/user.actions';
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
libs/shared/root/data-access/src/index.ts:36:1 - error TS2308: Module './lib/+state/auth/auth.actions' has already exported a member named 'UpdateSuccess'. Consider explicitly re-exporting to resolve the ambiguity.

36 export * from './lib/+state/user/user.actions';
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
libs/shared/root/data-access/src/index.ts:36:1 - error TS2308: Module './lib/+state/users/users.actions' has already exported a member named 'Find'. Consider explicitly re-exporting to resolve the ambiguity.

36 export * from './lib/+state/user/user.actions';
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
libs/shared/root/data-access/src/index.ts:36:1 - error TS2308: Module './lib/+state/users/users.actions' has already exported a member named 'FindSuccess'. Consider explicitly re-exporting to resolve the ambiguity.

36 export * from './lib/+state/user/user.actions';
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
libs/shared/root/data-access/src/index.ts:36:1 - error TS2308: Module './lib/+state/users/users.actions' has already exported a member named 'Reset'. Consider explicitly re-exporting to resolve the ambiguity.

36 export * from './lib/+state/user/user.actions';
```

### Environment
```
nx report

 >  NX   Report complete - copy this into the issue template

   Node : 16.15.1
   OS   : darwin x64
   yarn : 1.22.18
   
   nx : 14.3.6
   @nrwl/angular : 14.3.6
   @nrwl/cypress : 14.3.6
   @nrwl/detox : Not Found
   @nrwl/devkit : 14.3.6
   @nrwl/eslint-plugin-nx : 14.3.6
   @nrwl/express : 14.3.6
   @nrwl/jest : 14.3.6
   @nrwl/js : 14.3.6
   @nrwl/linter : 14.3.6
   @nrwl/nest : 14.3.6
   @nrwl/next : Not Found
   @nrwl/node : 14.3.6
   @nrwl/nx-cloud : 14.1.2
   @nrwl/nx-plugin : Not Found
   @nrwl/react : Not Found
   @nrwl/react-native : Not Found
   @nrwl/schematics : 8.12.11
   @nrwl/storybook : 14.3.6
   @nrwl/web : 14.3.6
   @nrwl/workspace : 14.3.6
   typescript : 4.7.4
   ---------------------------------------
   Community plugins:
   	 @fortawesome/angular-fontawesome: 0.11.1
   	 @ionic/angular: 6.1.10
   	 @ngrx/component-store: 14.0.0
   	 @ngrx/effects: 14.0.0
   	 @ngrx/entity: 14.0.0
   	 @ngrx/router-store: 14.0.0
   	 @ngrx/store: 14.0.0
   	 @nxtend/capacitor: 13.0.0
   	 @nxtend/ionic-angular: 13.1.0
   	 @storybook/angular: 6.5.9
   	 angular-builder-custom-terser-options: 1.0.1
   	 @ngrx/schematics: 14.0.0
   	 @ngrx/store-devtools: 14.0.0
   	 @nxpm/stack: 4.21.0
   	 @nxtend/capacitor: 13.0.0
   	 @storybook/angular: 6.5.9
```

--- a/packages/angular/src/generators/ngrx/files/classes-syntax/__directory__/__fileName__.facade.spec.ts__tmpl__
+++ b/packages/angular/src/generators/ngrx/files/classes-syntax/__directory__/__fileName__.facade.spec.ts__tmpl__
@@ -11,8 +11,8 @@ import { <%= className %>Facade } from './<%= fileName %>.facade';
 import {
   <%= className %>State,
   <%= className %>Entity,
-  initialState,
-  reducer
+  initial<%= className %>State,
+  <%= propertyName %>Reducer
 } from './<%= fileName %>.reducer';
 import { <%= propertyName %>Query } from './<%= fileName %>.selectors';
 
@@ -32,7 +32,7 @@ describe('<%= className %>Facade', () => {
     beforeEach(() => {
       @NgModule({
         imports: [
-          StoreModule.forFeature('<%= propertyName %>', reducer, { initialState }),
+          StoreModule.forFeature('<%= propertyName %>', <%= propertyName %>Reducer, { initialState: initial<%= className %>State }),
           EffectsModule.forFeature([<%= className %>Effects])
         ],
         providers: [<%= className %>Facade]

--- a/packages/angular/src/generators/ngrx/files/classes-syntax/__directory__/__fileName__.reducer.spec.ts__tmpl__
+++ b/packages/angular/src/generators/ngrx/files/classes-syntax/__directory__/__fileName__.reducer.spec.ts__tmpl__
@@ -1,5 +1,5 @@
 import { <%= className %>Loaded } from './<%= fileName %>.actions';
-import { <%= className %>State, <%= className %>Entity, initialState, reducer } from './<%= fileName %>.reducer';
+import { <%= className %>State, <%= className %>Entity, initial<%= className %>State, <%= propertyName %>Reducer } from './<%= fileName %>.reducer';
 
 describe('<%= className %> Reducer', () => {
   const get<%= className %>Id = (it: <%= className %>Entity) => it.id;
@@ -12,7 +12,7 @@ describe('<%= className %> Reducer', () => {
     it('should return the list of known <%= className %>', () => {
       const <%= propertyName %> = [create<%= className %>('PRODUCT-AAA'), create<%= className %>('PRODUCT-zzz')];
       const action = new <%= className %>Loaded(<%= propertyName %>);
-      const result: <%= className %>State = reducer(initialState, action);
+      const result: <%= className %>State = <%= propertyName %>Reducer(initial<%= className %>State, action);
       const selId: string = get<%= className %>Id(result.list[1]);
 
       expect(result.loaded).toBe(true);
@@ -24,9 +24,9 @@ describe('<%= className %> Reducer', () => {
   describe('unknown action', () => {
     it('should return the previous state', () => {
       const action = {} as any;
-      const result = reducer(initialState, action);
+      const result = <%= propertyName %>Reducer(initial<%= className %>State, action);
 
-      expect(result).toBe(initialState);
+      expect(result).toBe(initial<%= className %>State);
     });
   });
 });

--- a/packages/angular/src/generators/ngrx/files/classes-syntax/__directory__/__fileName__.reducer.ts__tmpl__
+++ b/packages/angular/src/generators/ngrx/files/classes-syntax/__directory__/__fileName__.reducer.ts__tmpl__
@@ -24,13 +24,13 @@ export interface <%= className %>PartialState {
   readonly [<%= constantName %>_FEATURE_KEY]: <%= className %>State;
 }
 
-export const initialState: <%= className %>State = {
+export const initial<%= className %>State: <%= className %>State = {
   list: [],
   loaded: false
 };
 
-export function reducer(
-  state: <%= className %>State = initialState, 
+export function <%= propertyName %>Reducer(
+  state: <%= className %>State = initial<%= className %>State,
   action: <%= className %>Action): <%= className %>State
 {
   switch (action.type) {

--- a/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.actions.ts__tmpl__
+++ b/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.actions.ts__tmpl__
@@ -1,7 +1,7 @@
 import { createAction, props } from '@ngrx/store';
 import { <%= className %>Entity } from './<%= fileName %>.models';
 
-export const init = createAction(
+export const init<%= className %> = createAction(
   '[<%= className %> Page] Init'
 );
 

--- a/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.effects.spec.ts__tmpl__
+++ b/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.effects.spec.ts__tmpl__
@@ -31,7 +31,7 @@ describe('<%= className %>Effects', () => {
 
   describe('init$', () => {
     it('should work', () => {
-      actions = hot('-a-|', { a: <%= className %>Actions.init() });
+      actions = hot('-a-|', { a: <%= className %>Actions.init<%= className %>() });
 
       const expected = hot('-a-|', { a: <%= className %>Actions.load<%= className %>Success({ <%= propertyName %>: [] }) });
 

--- a/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.effects.ts__tmpl__
+++ b/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.effects.ts__tmpl__
@@ -7,17 +7,17 @@ import * as <%= className %>Feature from './<%= fileName %>.reducer';
 
 @Injectable()
 export class <%= className %>Effects {
-  <% if (useDataPersistence) { %>init$ = createEffect(() => this.dataPersistence.fetch(<%= className %>Actions.init, {
-    run: (action: ReturnType<typeof <%= className %>Actions.init>, state: <%= className %>Feature.<%= className %>PartialState) => {
+  <% if (useDataPersistence) { %>init$ = createEffect(() => this.dataPersistence.fetch(<%= className %>Actions.<%= className %>init, {
+    run: (action: ReturnType<typeof <%= className %>Actions.init<%= className %>>, state: <%= className %>Feature.<%= className %>PartialState) => {
       // Your custom service 'load' logic goes here. For now just return a success action...
       return <%= className %>Actions.load<%= className %>Success({ <%= propertyName %>: [] });
     },
-    onError: (action: ReturnType<typeof <%= className %>Actions.init>, error) => {
+    onError: (action: ReturnType<typeof <%= className %>Actions.init<%= className %>>, error) => {
       console.error('Error', error);
       return <%= className %>Actions.load<%= className %>Failure({ error });
     }
   }));<% } else { %>init$ = createEffect(() => this.actions$.pipe(
-    ofType(<%= className %>Actions.init),
+    ofType(<%= className %>Actions.init<%= className %>),
     fetch({
       run: action => {
         // Your custom service 'load' logic goes here. For now just return a success action...

--- a/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.facade.spec.ts__tmpl__
+++ b/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.facade.spec.ts__tmpl__
@@ -11,14 +11,14 @@ import { <%= className %>Facade } from './<%= fileName %>.facade';
 import { <%= className %>Entity } from './<%= fileName %>.models';
 import {
   <%= constantName %>_FEATURE_KEY,
-  State,
-  initialState,
-  reducer
+  <%= className %>State,
+  initial<%= className %>State,
+  <%= propertyName %>Reducer
 } from './<%= fileName %>.reducer';
 import * as <%= className %>Selectors from './<%= fileName %>.selectors';
 
 interface TestSchema {
-  <%= propertyName %>: State;
+  <%= propertyName %>: <%= className %>State;
 }
 
 describe('<%= className %>Facade', () => {
@@ -33,7 +33,7 @@ describe('<%= className %>Facade', () => {
     beforeEach(() => {
       @NgModule({
         imports: [
-          StoreModule.forFeature(<%= constantName %>_FEATURE_KEY, reducer),
+          StoreModule.forFeature(<%= constantName %>_FEATURE_KEY, <%= propertyName %>Reducer),
           EffectsModule.forFeature([<%= className %>Effects])
         ],
         providers: [<%= className %>Facade]

--- a/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.facade.ts__tmpl__
+++ b/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.facade.ts__tmpl__
@@ -22,6 +22,6 @@ export class <%= className %>Facade {
    * or more tasks in your Effects.
    */
   init() {
-    this.store.dispatch(<%= className %>Actions.init());
+    this.store.dispatch(<%= className %>Actions.init<%= className %>());
   }
 }

--- a/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.reducer.spec.ts__tmpl__
+++ b/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.reducer.spec.ts__tmpl__
@@ -2,7 +2,7 @@ import { Action } from '@ngrx/store';
 
 import * as <%= className %>Actions from './<%= fileName %>.actions';
 import { <%= className %>Entity } from './<%= fileName %>.models';
-import { State, initialState, reducer } from './<%= fileName %>.reducer';
+import { <%= className %>State, initial<%= className %>State, <%= propertyName %>Reducer } from './<%= fileName %>.reducer';
 
 describe('<%= className %> Reducer', () => {
   const create<%= className %>Entity = (id: string, name = ''): <%= className %>Entity => ({
@@ -18,7 +18,7 @@ describe('<%= className %> Reducer', () => {
       ];
       const action = <%= className %>Actions.load<%= className %>Success({ <%= propertyName %> });
 
-      const result: State = reducer(initialState, action);
+      const result: <%= className %>State = <%= propertyName %>Reducer(initial<%= className %>State, action);
 
       expect(result.loaded).toBe(true);
       expect(result.ids.length).toBe(2);
@@ -29,9 +29,9 @@ describe('<%= className %> Reducer', () => {
     it('should return the previous state', () => {
       const action = {} as Action;
 
-      const result = reducer(initialState, action);
+      const result = <%= propertyName %>Reducer(initial<%= className %>State, action);
 
-      expect(result).toBe(initialState);
+      expect(result).toBe(initial<%= className %>State);
     });
   });
 });

--- a/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.reducer.ts__tmpl__
+++ b/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.reducer.ts__tmpl__
@@ -6,26 +6,26 @@ import { <%= className %>Entity } from './<%= fileName %>.models';
 
 export const <%= constantName %>_FEATURE_KEY = '<%= propertyName %>';
 
-export interface State extends EntityState<<%= className %>Entity> {
+export interface <%= className %>State extends EntityState<<%= className %>Entity> {
   selectedId?: string | number; // which <%= className %> record has been selected
   loaded: boolean; // has the <%= className %> list been loaded
   error?: string | null; // last known error (if any)
 }
 
 export interface <%= className %>PartialState {
-  readonly [<%= constantName %>_FEATURE_KEY]: State;
+  readonly [<%= constantName %>_FEATURE_KEY]: <%= className %>State;
 }
 
 export const <%= propertyName %>Adapter: EntityAdapter<<%= className %>Entity> = createEntityAdapter<<%= className %>Entity>();
 
-export const initialState: State = <%= propertyName %>Adapter.getInitialState({
+export const initial<%= className %>State: <%= className %>State = <%= propertyName %>Adapter.getInitialState({
   // set initial required properties
   loaded: false
 });
 
-const <%= propertyName %>Reducer = createReducer(
-  initialState,
-  on(<%= className %>Actions.init,
+const reducer = createReducer(
+  initial<%= className %>State,
+  on(<%= className %>Actions.init<%= className %>,
     state => ({ ...state, loaded: false, error: null })
   ),
   on(<%= className %>Actions.load<%= className %>Success,
@@ -36,6 +36,6 @@ const <%= propertyName %>Reducer = createReducer(
   ),
 );
 
-export function reducer(state: State | undefined, action: Action) {
-  return <%= propertyName %>Reducer(state, action);
+export function <%= propertyName %>Reducer(state: <%= className %>State | undefined, action: Action) {
+  return reducer(state, action);
 }

--- a/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.selectors.spec.ts__tmpl__
+++ b/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.selectors.spec.ts__tmpl__
@@ -1,5 +1,5 @@
 import { <%= className %>Entity } from './<%= fileName %>.models';
-import { <%= propertyName %>Adapter, <%= className %>PartialState, initialState } from './<%= fileName %>.reducer';
+import { <%= propertyName %>Adapter, <%= className %>PartialState, initial<%= className %>State } from './<%= fileName %>.reducer';
 import * as <%= className %>Selectors from './<%= fileName %>.selectors';
 
 describe('<%= className %> Selectors', () => {
@@ -19,7 +19,7 @@ describe('<%= className %> Selectors', () => {
         create<%= className %>Entity('PRODUCT-BBB'),
         create<%= className %>Entity('PRODUCT-CCC')
       ], {
-        ...initialState,
+        ...initial<%= className %>State,
         selectedId : 'PRODUCT-BBB',
         error: ERROR_MSG,
         loaded: true

--- a/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.selectors.ts__tmpl__
+++ b/packages/angular/src/generators/ngrx/files/creator-syntax/__directory__/__fileName__.selectors.ts__tmpl__
@@ -1,34 +1,34 @@
 import { createFeatureSelector, createSelector } from '@ngrx/store';
-import { <%= constantName %>_FEATURE_KEY, State, <%= propertyName %>Adapter } from './<%= fileName %>.reducer';
+import { <%= constantName %>_FEATURE_KEY, <%= className %>State, <%= propertyName %>Adapter } from './<%= fileName %>.reducer';
 
 // Lookup the '<%= className %>' feature state managed by NgRx
-export const get<%= className %>State = createFeatureSelector<State>(<%= constantName %>_FEATURE_KEY);
+export const get<%= className %>State = createFeatureSelector<<%= className %>State>(<%= constantName %>_FEATURE_KEY);
 
 const { selectAll, selectEntities } = <%= propertyName %>Adapter.getSelectors();
 
 export const get<%= className %>Loaded = createSelector(
   get<%= className %>State,
-  (state: State) => state.loaded
+  (state: <%= className %>State) => state.loaded
 );
 
 export const get<%= className %>Error = createSelector(
   get<%= className %>State,
-  (state: State) => state.error
+  (state: <%= className %>State) => state.error
 );
 
 export const getAll<%= className %> = createSelector(
   get<%= className %>State,
-  (state: State) => selectAll(state)
+  (state: <%= className %>State) => selectAll(state)
 );
 
 export const get<%= className %>Entities = createSelector(
   get<%= className %>State,
-  (state: State) => selectEntities(state)
+  (state: <%= className %>State) => selectEntities(state)
 );
 
 export const getSelectedId = createSelector(
   get<%= className %>State,
-  (state: State) => state.selectedId
+  (state: <%= className %>State) => state.selectedId
 );
 
 export const getSelected = createSelector(

--- a/packages/angular/src/generators/ngrx/lib/add-imports-to-module.ts
+++ b/packages/angular/src/generators/ngrx/lib/add-imports-to-module.ts
@@ -47,6 +47,7 @@ export function addImportsToModule(
   const effectsName = `${names(options.name).className}Effects`;
   const facadeName = `${names(options.name).className}Facade`;
   const className = `${names(options.name).className}`;
+  const propertyName = `${names(options.name).propertyName}`;
   const reducerImports = `* as from${className}`;
 
   const storeMetaReducers = `metaReducers: !environment.production ? [] : []`;
@@ -61,7 +62,7 @@ export function addImportsToModule(
   const nxModule = 'NxModule.forRoot()';
   const effectsForRoot = `EffectsModule.forRoot([${effectsName}])`;
   const effectsForEmptyRoot = `EffectsModule.forRoot([])`;
-  const storeForFeature = `StoreModule.forFeature(from${className}.${constantName}_FEATURE_KEY, from${className}.reducer)`;
+  const storeForFeature = `StoreModule.forFeature(from${className}.${constantName}_FEATURE_KEY, from${className}.${propertyName}Reducer)`;
   const effectsForFeature = `EffectsModule.forFeature([${effectsName}])`;
   const devTools = `!environment.production ? StoreDevtoolsModule.instrument() : []`;
   const storeRouterModule = 'StoreRouterConnectingModule.forRoot()';

